### PR TITLE
feat(auth/cognito): base directory structure w/ signUp API

### PR DESCRIPTION
**Pull request by haverchuck**
#### Description of changes
This PR is meant to establish patterns and conventions for all subseqent vNext API PRs. These patterns include:
* directory structures for APIs, types, utils and sdk clients.
* a pattern for creating abstractions over SDK clients, such that the clients themselves can be swapped out without requiring changes within customer-facing API code.
* unit test patterns
* documentation
* standardized changes to package.jsons to ensure proper type definition generations/access.

The general directory structure being used is:
```
auth
│   package.json
│   [other root files]
│
└───__tests__
│   └───providers
│       └───cognito
│       │   signUp.test.ts
└───src
│   └───providers
│       └───cognito
│           └───apis
            | signUp.ts
│           └───types. <-- provider level types
│               └───models
│               └───options
│          └───utils
│   └───types.  <-- category level types
│               └───enums
│               └───models
│               └───options
│               └───requests
│               └───results
```

Additionally, this PR includes a nearly-complete implementation of the Cognito Plugin SignUp API. It will need to be updated with autoSignIn, error handling, and additional unit tests. Furthermore, the signUp API may need additional API BR review.

Please use the discussion in the PR to flesh out these patterns. The goal here is for every subsequent PR to use similar patterns and conventions, so we don't have to re-hash the same conversations across multiple simultaneous PRs.

#### Description of how you validated changes
`npx jest signUp.test.ts`

#### Checklist

- [X] PR description included
- [X] `yarn test` passes
- [ X Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/__tests__/providers/cognito/signUp.test.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import { AuthSignUpStep, signUp } from '../../..';
+import {
+	SignUpClientInput,
+	UserpoolClient,
+} from '../../../lib/providers/cognito/utils/clients/UserPoolClient';
+import { testParams } from './testUtils/testParams';
+
+let signUpSpy;
+
+describe('SignUp API Happy Path Cases:', () => {
```
Comment (haverchuck) [inline packages/auth/__tests__/providers/cognito/signUp.test.ts:14]: My thought is that every API unit test file will have three describe blocks: Happy Path, Error Case, and Edge Case. If we need to subdividie into more complexity we can do so, but in those case we can create these three describe blocks within larger ones.

The goal here is make our unit test files easier to navigate and maintain.

File: packages/auth/__tests__/providers/cognito/signUp.test.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { SignUpCommandOutput } from '@aws-sdk/client-cognito-identity-provider';
+import { AuthSignUpStep, signUp } from '../../..';
+import {
+	SignUpClientInput,
+	UserpoolClient,
+} from '../../../lib/providers/cognito/utils/clients/UserPoolClient';
+import { testParams } from './testUtils/testParams';
```
Comment (haverchuck) [inline packages/auth/__tests__/providers/cognito/signUp.test.ts:10]: In order to reduce the number of imports at the top of our test files, I've created one large "testParams" object. I can see arguments for and against this.

File: packages/auth/package.json [hunk 3]
```diff
@@ -47,6 +69,8 @@
   ],
   "dependencies": {
     "@aws-amplify/core": "5.0.15",
+    "@aws-sdk/client-cognito-identity-provider": "3.54.0",
```
Comment (haverchuck) [inline packages/auth/package.json:72]: These sdks are being imported only for Input/Output types, not clients. We haven't fully decided on this approach yet.

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import { UserPoolHttpClient } from './HttpClients';
+
+export type SignUpClientInput = Partial<SignUpCommandInput> &
+	Pick<SignUpCommandInput, 'Username'> &
+	Pick<SignUpCommandInput, 'Password'> &
+	Pick<SignUpCommandInput, 'UserAttributes'> &
+	Pick<SignUpCommandInput, 'ClientMetadata'> &
+	Pick<SignUpCommandInput, 'ValidationData'>;
+
+export class UserpoolClient {
```
Comment (haverchuck) [inline packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:18]: This class (consisting of static methods) is what every individual API will interact with. In this case, it is in turn using the http client that @elorzafe made with the POC. The goal of this class is to allow for us to easily swap out temporary clients for sdk clients when they are ready, without having to touch all of our implementation code.

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+	SignUpCommandInput,
+	SignUpCommandOutput,
+} from '@aws-sdk/client-cognito-identity-provider';
+import { Amplify } from '@aws-amplify/core';
+import { UserPoolHttpClient } from './HttpClients';
+
+export type SignUpClientInput = Partial<SignUpCommandInput> &
```
Comment (haverchuck) [inline packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:11]: Using partials so we don't have to manually pass things like `ClientID`.

File: packages/auth/src/providers/cognito/utils/clients/HttpClients.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+			...this._headers,
+			'X-Amz-Target': `AWSCognitoIdentityProviderService.${operation}`,
+		};
+		const options: RequestInit = {
+			headers,
+			method: 'POST',
+			mode: 'cors',
+			body: JSON.stringify(input),
+		};
+		return (await (await fetch(this._endpoint, options)).json()) as T;
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/utils/clients/HttpClients.ts:42]: I know error handling is out of scope for now, but just food for thought I wonder if we should make Type guards for responses & give us a place to validate the response while asserting the type.

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts (hunk 4 above)
Comment (jimblanc) [reply packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:18]: IMO we'll still probably need to do some integration work here when the time comes, e.g. depending on the decision to AWS SDK types in the public interface, etc. Approach is good to unblock work though.

File: packages/core/src/Amplify.ts [hunk 7]
```diff
@@ -5,6 +5,7 @@ const logger = new LoggerClass('Amplify');
 export class AmplifyClass {
 	// Everything that is `register`ed is tracked here
 	private _components = [];
+
```
Comment (jimblanc) [inline packages/core/src/Amplify.ts:8]:
```suggestion
```

**Review by jimblanc — COMMENTED** [review]
Thanks for putting this together Dustin!

Comments on the description:
- Should include the provider index file in the file list, which will be used to generate provider types.
- I wonder how applicable some of these category-level API types will be to non-Cognito providers. We can always tweak if needed, but backwards compatibility might be a pain depending on when we need to do it. I'm wondering if it makes sense to have category-level API definitions vs just treating providers independently?

One thing this doesn't touch on is the umbrella package (which is covered by a doc Francisco & I are working on), but it's probably okay to leave that out for now until we have a clear direction on the scoped packages.

File: packages/auth/src/providers/cognito/apis/signUp.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+				},
+			},
+		};
+	}
+}
+
+function mapValidationData(data: ValidationData): AttributeType[] {
+	const attributeTypeArray: AttributeType[] = [];
+	if (data) {
+		for (let k in data) {
```
[the remark this answers is not in the document]
Comment (helgabalashova) [reply packages/auth/src/providers/cognito/apis/signUp.ts:86]:
```
Object.entries(data).map(([key, value]) => ({
				Name: key,
				Value: value
}));
```

File: packages/auth/src/types/enums/AuthSignUpStep.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * Denotes the next step in the Sign Up process.
+ */
+export const enum AuthSignUpStep {
```
Comment (israx) [inline packages/auth/src/types/enums/AuthSignUpStep.ts:7]: I think something that I forgot to update was removing the `const` keyword for enums that are public. 

`const` enums are beneficial for bundle size reduction but are not grate for public use.

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts (hunk 4 above)
Comment (haverchuck) [reply packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:18]: Yeah - there will def be implementation work, but at least it will all be in one place.

File: packages/auth/package.json (hunk 3 above)
Comment (haverchuck) [reply packages/auth/package.json:72]: @jimblanc I don't think so - I tried that but they couldn't be found.

File: packages/core/src/Amplify.ts [hunk 10]
```diff
@@ -32,6 +32,10 @@ export class AmplifyClass {
 	Logger = LoggerClass;
 	ServiceWorker = null;
 
+	public get config(): any {
+		return this._config;
```
Comment (elorzafe) [inline packages/core/src/Amplify.ts:36]: I prefer to return a copy just to avoid issues of modifying the `this._config` outside of the Singleton

File: packages/auth/src/providers/cognito/types/options/CognitoSignUpOptions.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+
+import { ClientMetadata, ValidationData } from '../..';
+
+/**
+ * Options specific to a Cognito Sign Up request.
+ */
+export type CognitoSignUpOptions = {
+	validationData?: ValidationData;
+	clientMetadata?: ClientMetadata;
+	// autoSignIn?: AutoSignInOptions;
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/types/options/CognitoSignUpOptions.ts:12]: I would add a TODO here

File: packages/auth/package.json (hunk 3 above)
Comment (elorzafe) [reply packages/auth/package.json:72]: Did you run `yarn && yarn bootstrap` after changing to devDependencies?

File: packages/auth/__tests__/providers/cognito/testUtils/testParams.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+export const testParams = {
```
Comment (elorzafe) [inline packages/auth/__tests__/providers/cognito/testUtils/testParams.ts:4]: I would be more explicit with the name here. What is the purpose of this `testParams`?

File: packages/auth/__tests__/providers/cognito/signUp.test.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+let signUpSpy;
```
Comment (elorzafe) [inline packages/auth/__tests__/providers/cognito/signUp.test.ts:12]: do you need this outside of the `describe`?

File: packages/auth/src/providers/cognito/apis/signUp.ts [hunk 14]
```diff
@@ -0,0 +1,89 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Amplify } from '@aws-amplify/core';
+import {
+	AttributeType,
+	SignUpCommandOutput,
+} from '@aws-sdk/client-cognito-identity-provider';
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/apis/signUp.ts:8]: can we do `import type { .... }`?

File: packages/auth/src/providers/cognito/apis/signUp.ts [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+): Promise<AuthSignUpResult<AuthStandardAttributeKey | CustomAttribute>> {
+	let validationData: AttributeType[] | undefined;
+	const _config = Amplify.config;
+	if (signUpRequest.options?.serviceOptions?.validationData) {
+		validationData = mapValidationData(
+			signUpRequest.options?.serviceOptions?.validationData
+		);
+	}
+
+	const res: SignUpCommandOutput = await UserpoolClient.signUp({
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/apis/signUp.ts:42]: Are we going to add error handling here? if is that the case I would add a `TODO` just for remind where it should be

File: packages/auth/src/providers/cognito/utils/clients/HttpClients.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+	SignUpCommandInput,
+	ConfirmSignUpCommandInput,
+	SignUpCommandOutput,
+} from '@aws-sdk/client-cognito-identity-provider';
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/utils/clients/HttpClients.ts:8]: can we do `import type { ... } `?

File: packages/auth/src/providers/cognito/utils/clients/HttpClients.ts (hunk 6 above)
Comment (elorzafe) [reply packages/auth/src/providers/cognito/utils/clients/HttpClients.ts:42]: I agree with Jim, at least where that logic should be, not need to implement that on this PR

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts [hunk 17]
```diff
@@ -0,0 +1,36 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+	SignUpCommandInput,
+	SignUpCommandOutput,
+} from '@aws-sdk/client-cognito-identity-provider';
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:7]: can we do `import type {...}`?

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts (hunk 4 above)
Comment (elorzafe) [reply packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:18]: Why does it need to be a class with static methods instead of just the methods?

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+export class UserpoolClient {
+	public static async signUp(
+		params: SignUpClientInput
+	): Promise<SignUpCommandOutput> {
+		const config = Amplify.config;
+		// TODO: update when config is typed properly
+		const region = config['aws_cognito_region'];
+		const clientId = config['aws_user_pools_web_client_id'];
+		const client = new UserPoolHttpClient(region);
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:26]: IMHO it would be more cleaner if this param are not part of each function, I think this code will be same on each function. What do you think?

**Review by elorzafe — COMMENTED** [review]
Minor comments, overall looks good

File: packages/auth/src/providers/cognito/types/options/CognitoSignUpOptions.ts (hunk 11 above)
Comment (haverchuck) [reply packages/auth/src/providers/cognito/types/options/CognitoSignUpOptions.ts:12]: I made one in the signUp API implementation.

File: packages/auth/__tests__/providers/cognito/testUtils/testParams.ts (hunk 12 above)
Comment (haverchuck) [reply packages/auth/__tests__/providers/cognito/testUtils/testParams.ts:4]: Changed to `authAPITestParams`, and used destructure in the test to avoid the verbosity.

File: packages/auth/__tests__/providers/cognito/signUp.test.ts (hunk 13 above)
Comment (haverchuck) [reply packages/auth/__tests__/providers/cognito/signUp.test.ts:12]: It might be handy if we want to manipulate across describe blocks, but for now I've moved it.

File: packages/auth/src/providers/cognito/apis/signUp.ts (hunk 15 above)
Comment (haverchuck) [reply packages/auth/src/providers/cognito/apis/signUp.ts:42]: Yeah - in the PR description I say that this doesn't include error handling. I think we need to align on that as a team. I didn't want to rely on Todos for that in these files, because we will need it at several junctures.

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts (hunk 18 above)
Comment (haverchuck) [reply packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:26]: Yeah, good call.

Comment (haverchuck) [reply packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:26]: Good call - I changed it.

**Comment by haverchuck** [comment]
@elorzafe 
> Why does it need to be a class with static methods instead of just the methods?
Initially I was trying to avoid instantiating a client every time, but hoisting the config stuff outside of each method makes instantiating worth it, so I removed the static methods.

File: packages/auth/package.json [hunk 19]
```diff
@@ -30,6 +30,28 @@
...
     "ts-coverage": "typescript-coverage-report -p ./tsconfig.build.json -t 77.44"
   },
+  "typesVersions": {
+    ">=3.8": {
+      "*": [
+        "./lib-esm/index.d.ts"
+      ],
+      "cognito": [
+        "./lib-esm/providers/cognito/index.d.ts"
+      ]
```
Comment (AllanZhengYP) [inline packages/auth/package.json:40]: I'm not sure this is the right syntax. The math mapping should exist in the tsconfig paths config IMO

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+} from '@aws-sdk/client-cognito-identity-provider';
+import { Amplify } from '@aws-amplify/core';
+import { UserPoolHttpClient } from './HttpClients';
+
+export type SignUpClientInput = Partial<SignUpCommandInput> &
+	Pick<SignUpCommandInput, 'Username'> &
+	Pick<SignUpCommandInput, 'Password'> &
+	Pick<SignUpCommandInput, 'UserAttributes'> &
+	Pick<SignUpCommandInput, 'ClientMetadata'> &
+	Pick<SignUpCommandInput, 'ValidationData'>;
```
Comment (AllanZhengYP) [inline packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:16]: Very nit: this can be translated to `Pick<SignUpCommandInput, 'Username'|'Password'|'UserAttributes'|'ClientMetadata'|'ValidationData'>`

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts (hunk 4 above)
Comment (AllanZhengYP) [reply packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:18]: With the proposed custom SDK signature, this can be updated to something like:
```
public async signUp(params: SignUpClientInput): Promise<SignUpCommandOutput> {
    return await signUp({ this.region }, {
				...params,
				ClientId: this.clientId,
    });
}
```

File: packages/auth/src/providers/cognito/utils/clients/HttpClients.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+const USER_AGENT = 'amplify test';
+
+export type ClientInputs = SignUpCommandInput;
+export type ClientOutputs = SignUpCommandOutput;
+export type ClientOperations = 'SignUp' | 'ConfirmSignUp';
+
+export class UserPoolHttpClient {
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/auth/src/providers/cognito/utils/clients/HttpClients.ts:15]: Per the latest revison, this should live in the provider now.

File: packages/auth/src/providers/cognito/types/models/ClientMetadata.ts
Comment (AllanZhengYP) [inline packages/auth/src/providers/cognito/types/models/ClientMetadata.ts:1]: What's the guideline of files going in the `models` folder?

File: packages/auth/package.json (hunk 19 above)
Comment (haverchuck) [reply packages/auth/package.json:40]: @AllanZhengYP I worked with @jimblanc on this - it's what worked on the POC I think. What would the alternative actually look like?

File: packages/auth/src/providers/cognito/types/models/ClientMetadata.ts
Comment (haverchuck) [reply packages/auth/src/providers/cognito/types/models/ClientMetadata.ts:1]: It's not SUPER prescriptive, but generally anything representing a granular type that makes up larger types such as options, requests or responses.

File: packages/auth/package.json (hunk 19 above)
Comment (AllanZhengYP) [reply packages/auth/package.json:40]: I was a little concern because it's not suggested or mentioned in the [TypeScript doc](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#version-selection-with-typesversions). But I think this will work after more thoughts, and reading the [SO answer](https://stackoverflow.com/questions/73149401/how-to-correctly-use-exports-and-subpaths-on-package-json-using-typescript#answer-75098673)

**Review by AllanZhengYP — DISMISSED** [review]

**Review by AllanZhengYP — APPROVED** [review]

File: packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { Amplify } from '@aws-amplify/core';
+
+export const UserPoolClient = {
+	// TODO: update when config is typed properly
+	region: Amplify.config['aws_cognito_region'],
+	clientId: Amplify.config['aws_user_pools_web_client_id'],
+};
```
Comment (elorzafe) [inline packages/auth/src/providers/cognito/utils/clients/UserPoolClient.ts:9]: This might need to be a function that returns this otherwise this might be static to what value was obtained when the module was imported. I am fine to add a comment about this on the code

**Review by elorzafe — APPROVED** [review]
I approve it but I think is necessary to mention on the `UserPoolClient` that needs to be refactored about how is going to read the configuration from `Amplify`

**Review by jimblanc — APPROVED** [review]

File: packages/auth/package.json (hunk 19 above)
Comment (jimblanc) [reply packages/auth/package.json:40]: Yeah the reason we did this was because the `types` field in `exports` doesn't actually do anything
